File: Algorithms_&_Problem_Solving_Patterns/shellsort.js
function shellSort(unsorted_array){

    let length_of_array = unsorted_array.length;

    // showing the array before sorting
    console.log(unsorted_array);

    for (let gap = Math.floor(length_of_array/2); gap > 0; gap = gap/2) {
        
        for (let i = gap; i < length_of_array; i++) {
            
            let temp = unsorted_array[i];
            let j;

            for (j = i; j>=gap && unsorted_array[j-gap] > unsorted_array[j]; j = j-gap){
                
                // swapping the elements if they are unsorted
                unsorted_array[j] = unsorted_array[j-gap];
                unsorted_array[j-gap] = temp;
            }
            
        }
        
    }

    // CAUTION 🚨🚨 because now the unsorted array not remained unsorted
    // Don't go with the name of array 🤣🤣🤣
    
    console.log(unsorted_array);
}

shellSort([12,34,54,2,3])

File: Algorithms_&_Problem_Solving_Patterns/shellsort.cpp
#include<bits/stdc++.h>
using namespace std;

int shellSort(int arr[],int n){
    for(int gap = n/2;gap>0;gap /= 2){
        
        for(int i = gap;i<n;i++){

            int temp = arr[i];
            int j;

            for(j = i;j>=gap && arr[j-gap] > temp; j -= gap){
                arr[j] = arr[j-gap];
                arr[j-gap] = temp;
            }
        }
    }
}

void printArray(int arr[],int n){
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}

int main(){
    int arr[] = {12,34,54,2,3};
    int n = sizeof(arr)/sizeof(arr[0]);
    cout<<"the array before sorting"<<endl;
    printArray(arr,n);
    shellSort(arr,n);
    cout<<"The array after sorting"<<endl;
    printArray(arr,n);
}